### PR TITLE
docs: add metadata documentation to httpClient

### PR DESCRIPTION
**Pull request by kapunahelewong**
Add documentation for http metadata.

### CODE REVIEW

File: aio/content/examples/http/src/app/retry-interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// #enddocregion reading-context
+
+// #docregion context-token
+export const ERROR_COUNT = new HttpContextToken(() => 0);
+// #endocregion context-token
+
+// #docregion set-context
+this.http.get('/data/feed', {
+  context: new HttpContext().set(RETRY_COUNT, 5),
+}).subscribe(results => {...});
```
Comment (petebacondarwin) [inline aio/content/examples/http/src/app/retry-interceptor.ts:12]: This docregion uses the `this` reference outside of a class, which is not really valid.
Instead I think we need to put this inside a class (perhaps a dummy service), which has the HttpClient injected into it.

File: aio/content/examples/http/src/app/retry-interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// #endocregion set-context
+
+
+// #docregion mutable-context
+export class RetryInterceptor implements HttpInterceptor {
+  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    const retryCount = req.context.get(RETRY_COUNT
```
Comment (petebacondarwin) [inline aio/content/examples/http/src/app/retry-interceptor.ts:19]:
```suggestion
    const retryCount = req.context.get(RETRY_COUNT);
```

File: aio/content/guide/http.md [hunk 3]
```diff
@@ -1196,3 +1196,50 @@ Alternatively, you can call `request.error()` with an `ErrorEvent`.
...
+### Setting context values when making a request
+
+When making a request, you can provide an `HttpContext` instance, in which you have already set the context values.
+
+<code-example path="http/src/app/retry-interceptors.ts" region="set-context" header="setting context values"></code-example>
+
+### Reading context values in an interceptor
+
+Within an interceptor, you can read the value of a token in a given request's context with `HttpContext.get()`.
+If you have not explicitly set a value for the token, Angular returns its default value.
```
Comment (petebacondarwin) [inline aio/content/guide/http.md:1235]:
```suggestion
If you have not explicitly set a value for the token, Angular returns the default value specified in the the token.
```

File: aio/content/guide/http.md [hunk 4]
```diff
@@ -1196,3 +1196,50 @@ Alternatively, you can call `request.error()` with an `ErrorEvent`.
...
+
+<code-example path="http/src/app/retry-interceptors.ts" region="reading-context" header="reading context values in an interceptor"></code-example>
```
Comment (petebacondarwin) [inline aio/content/guide/http.md:1237]: I think the reading-context docregion is not setup correctly... it only contains the `import` statement.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Thanks for putting this together @kapunahelewong - I have a couple of suggestions.

File: aio/content/examples/http/src/app/retry-interceptor.ts (hunk 1 above)
Comment (kapunahelewong) [reply aio/content/examples/http/src/app/retry-interceptor.ts:12]: Like this, @petebacondarwin?

```
export class FakeService {
  constructor(private httpClient: HttpClient) {}
  // #docregion set-context
  this.http.get('/data/feed', {
    context: new HttpContext().set(RETRY_COUNT, 5),
  }).subscribe(results => {...});
  // #endocregion set-context
}
```

File: aio/content/guide/http.md [hunk 5]
```diff
@@ -1196,3 +1196,50 @@ Alternatively, you can call `request.error()` with an `ErrorEvent`.
...
+By default, the interceptor might retry a request three times, but you might wish to override this retry count for particularly error-prone or sensitive requests.
+
+`HttpClient` requests contain a _context_ which can carry metadata about the request.
+This context is available for interceptors to read or modify, though it is not transmitted to the backend server when the request is sent.
+This allows applications or other interceptors to tag requests with configuration parameters, such as how many times to retry a request.
+
+### Creating a context token
+
+Angular stores and retrieves a value in the context using an `HttpContextToken`.
+You can create a context token using the `new` operator as in the following example:
```
Comment (TeriGlover) [inline aio/content/guide/http.md:1213]:
```suggestion
You can create a context token using the `new` operator, as in the following example:
```

File: aio/content/guide/http.md [hunk 6]
```diff
@@ -1196,3 +1196,50 @@ Alternatively, you can call `request.error()` with an `ErrorEvent`.
...
+Within an interceptor, you can read the value of a token in a given request's context with `HttpContext.get()`.
+If you have not explicitly set a value for the token, Angular returns the default value specified in the token.
+
+<code-example path="http/src/app/retry-interceptors.ts" region="reading-context" header="reading context values in an interceptor"></code-example>
+
+### Contexts are mutable
+
+Unlike most other aspects of `HttpRequest` instances, the request context is mutable and persists across other immutable transformations of the request.
+This allows interceptors to coordinate operations through the context.
+For example, the `RetryInterceptor` example could use a second context token to track how many errors occur during the execution of a given request:
```
Comment (TeriGlover) [inline aio/content/guide/http.md:1243]:
```suggestion
For example, the `RetryInterceptor` example could use a second context token to track how many errors occur during the execution of a given request:
```
Suggestions to replace "example" mentioned twice so close together:
For instance, ...
To illustrate, ...
For example,`RetryInterceptor` could use...

**Review by TeriGlover — APPROVED** [review]
A few suggestions; otherwise, LGTM.

File: aio/content/examples/http/src/app/retry-interceptor.ts (hunk 1 above)
Comment (petebacondarwin) [reply aio/content/examples/http/src/app/retry-interceptor.ts]: Yeah looks good!

File: aio/content/guide/http.md [hunk 7]
```diff
@@ -1196,3 +1196,50 @@ Alternatively, you can call `request.error()` with an `ErrorEvent`.
...
   region="network-error">
 </code-example>
+
+
+## Passing metadata to interceptors
+
+Many interceptors require or benefit from configuration. Consider an interceptor which retries failed requests.
+By default, the interceptor might retry a request three times, but you might wish to override this retry count for particularly error-prone or sensitive requests.
+
+`HttpClient` requests contain a _context_ which can carry metadata about the request.
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply aio/content/guide/http.md:1206]: I would say "that" in this case.

**Review by aikithoughts — APPROVED** [review]
Reviewed-for: global-docs-approvers

**Comment by petebacondarwin** [comment]
@googlebot I consent.

**Review by petebacondarwin — APPROVED** [review]

**Pull request closed without merging**